File: Chapter11/VB/SystemBrowser/Coordinators/CoordinatorLifecycle.vb
Public Class LifecycleCoordinator

  ' The main entry point for the application
  Shared Sub Main()
    Dim coordinator As New LifecycleCoordinator
    coordinator.Start() ' app exits when this call completes
  End Sub

  Enum LifecycleState
    StartingUp
    Running
    ShuttingDown
  End Enum

  Private state As LifecycleState

  ' shortcut property
  ReadOnly Property Builder() As Builder
    Get
      Return Builder.Singleton
    End Get
  End Property

  Public Sub Start()
    StartupSystem()
    RunSystem() ' app exits when this call completes
  End Sub

  Sub StartupSystem()
    state = LifecycleState.StartingUp
    Builder._startup.Run()
  End Sub

  Sub RunSystem()
    If state <> LifecycleState.StartingUp Then
      Throw New Exception("Invalid lifecycle state")
    End If

    state = LifecycleState.Running
    AddHandler Builder._cruise.OnExiting, AddressOf ShutDownSystem
    Builder._cruise.Run() ' app exits when this call completes
  End Sub

  Sub ShutDownSystem()
    If state <> LifecycleState.Running Then
      Throw New Exception("Invalid lifecycle state")
    End If

    state = LifecycleState.ShuttingDown
    Builder._shutdown.Run()
  End Sub

End Class

File: Chapter11/C#/SystemBrowser/Coordinators/CoordinatorLifecycle.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

namespace SystemBrowser
{
  public class LifecycleCoordinator
  {
    // The main entry point for the application
    [STAThread]
    static void Main() 
    {
      LifecycleCoordinator coordinator = new LifecycleCoordinator();
      coordinator.Start(); // app exits when this call completes
    }
    
    enum LifecycleState {StartingUp, Running, ShuttingDown};
    LifecycleState state;

    // shortcut property
    Builder Builder
    {
      get {return Builder.Singleton;}
    }

    public void Start()
    {
      StartupSystem();
      RunSystem(); // app exits when this call completes
    }
    
    void StartupSystem()  
    {
      state = LifecycleState.StartingUp;
      Builder.startup.Run();
    }
    
    void RunSystem()
    {
      if (state != LifecycleState.StartingUp)
        throw new Exception("Invalid lifecycle state");

      state = LifecycleState.Running;
      Builder.cruise.OnExiting += new CoordinatorCruise.ExitingHandler(ShutDownSystem);
      Builder.cruise.Run(); // app exits when this call completes
    }

    void ShutDownSystem()
    {
      if (state != LifecycleState.Running)
        throw new Exception("Invalid lifecycle state");

      state = LifecycleState.ShuttingDown;
      Builder.shutdown.Run();
    }
  }
}
